Forge.TreeWalker/src/TreeWalkerParameters.cs

```diff
@@ -180,8 +188,89 @@ public class TreeWalkerParameters
+        public TreeWalkerParameters(
+            Guid sessionId,
+            ForgeTree forgeTree,
+            IForgeDictionary forgeState,
+            ITreeWalkerCallbacksV2 callbacksV2,
+            CancellationToken token)
+        {
+            if (sessionId == Guid.Empty) throw new ArgumentNullException("sessionId");
+            if (forgeTree == null) throw new ArgumentNullException("forgeTree");
+            if (forgeState == null) throw new ArgumentNullException("forgeState");
+            if (callbacksV2 == null) throw new ArgumentNullException("callbacksV2");
+            if (token == null) throw new ArgumentNullException("token");
+
+            this.SessionId = sessionId;
+            this.ForgeTree = forgeTree;
+            this.ForgeState = forgeState;
+            this.CallbacksV2 = callbacksV2;
+            this.Token = token;
+        }
+
+        /// <summary>
+        /// Instantiates a TreeWalkerParameters object with the properties that are required to instantiate a TreeWalkerSession object.
+        /// Note: Recommended to cache the deserialized ForgeTree in your app and use that to avoid Forge deserializing the JsonSchema each initialize.
+        /// </summary>
+        /// <param name="sessionId">The unique identifier for this session.</param>
+        /// <param name="jsonSchema">The JSON schema.</param>
+        /// <param name="forgeState">The Forge state.</param>
+        /// <param name="callbacksV2">The callbacks object.</param>
+        /// <param name="token">The cancellation token.</param>
+        public TreeWalkerParameters(
+            Guid sessionId,
+            string jsonSchema,
+            IForgeDictionary forgeState,
+            ITreeWalkerCallbacksV2 callbacksV2,
+            CancellationToken token)
+        {
+            if (sessionId == Guid.Empty) throw new ArgumentNullException("sessionId");
+            if (string.IsNullOrWhiteSpace(jsonSchema)) throw new ArgumentNullException("jsonSchema");
+            if (forgeState == null) throw new ArgumentNullException("forgeState");
+            if (callbacksV2 == null) throw new ArgumentNullException("callbacksV2");
+            if (token == null) throw new ArgumentNullException("token");
+
+            this.SessionId = sessionId;
+            this.JsonSchema = jsonSchema;
+            this.ForgeState = forgeState;
+            this.CallbacksV2 = callbacksV2;
+            this.Token = token;
+        }
+        #endregion Constructor with ITreeWalkerCallbacksV2
+
+        private void TryUpgradeToITreeWalkerCallbacksV2(ITreeWalkerCallbacks callbacks)
+        {
+            if (callbacks is ITreeWalkerCallbacksV2)
+            {
+                ITreeWalkerCallbacksV2 callbacksV2 = callbacks as ITreeWalkerCallbacksV2;
+                if (callbacksV2 != null)
+                {
+                    this.CallbacksV2 = callbacksV2;
+                    return;
+                }
+            }
+
+            this.Callbacks = callbacks;
```

Comment:
>this.Callbacks = callbacks [](start = 12, length = 26)

Here's another idea for implementing the V2 backwards compatibility logic:

Keep only the public property ITreeWalkerCallbacks.
Have a private bool that is true if Callbacks is castable to V2.
As an optimization to save ourselves from casting Callbacks to V2 each time, also cache a private ITreeWalkerCallbacksV2 property here, and use that when bool is true.
Retrieve these new private values either through Get methods or perhaps just make them public with private set.

The benefits of doing it this way:
No additional public properties to worry about.
No additional constructors to worry about.
V2 is automatically taken care of under the covers.

Reply:
Yes, you list another option to do it:

No additional public properties to worry about. // yes, no additional public. but need a new Get function to get the V2.

No additional constructors to worry about.  // yes, users can keep use the existing constructor without breaking anything. On the second thought, our purpose is to let them switch to the new interface as early as possible. If they keep use the V1 interface, then we have to keep the if-else logic in TreeWalkerSession.WalkTree(), which is we want to avoid. Leave the V1 interface is kind of encourage the opposite in the long run.

V2 is automatically taken care of under the covers. // Same reason as above. I would like to convert all users to the new V2 Interface and only support V1 under the covers as a fallback plan.

Reply:
Yes, you list another option to do it:

No additional public properties to worry about. // yes, no additional public. but need a new Get function to get the V2.

No additional constructors to worry about.  // yes, users can keep use the existing constructor without breaking anything. On the second thought, our purpose is to let them switch to the new interface as early as possible. If they keep use the V1 interface, then we have to keep the if-else logic in TreeWalkerSession.WalkTree(), which is we want to avoid. Leave the V1 interface is kind of encourage the opposite in the long run.

V2 is automatically taken care of under the covers. // Same reason as above. I would like to convert all users to the new V2 Interface and only support V1 under the covers as a fallback plan.

---
In reply to: [504228085](https://github.com/microsoft/Forge/pull/32#discussion_r504228085) [](ancestors = 504228085)

Reply:
We cannot know when users switch to V2 because we can't tell which version folks are using. The transition will naturally happen over time when folks want to use new features that are in V2.
I'm not sure which implementation to go with, they all seem fine. We want the implementation to be as smooth as possible when we introduce V2, when we eventually deprecate V1, and with least duplicate code.

---
In reply to: [504278218](https://github.com/microsoft/Forge/pull/32#discussion_r504278218) [](ancestors = 504278218,504228085)

Reply:
I agree with you here about having the V2 constructors. We want folks to use them so eventually we can remove the V1 constructors and interface.

Because of this, let's force folks to use the V2 constructor if they wish to take advantage of the V2 features. We do this by removing TryUpgradeToITreeWalkerCallbacksV2. Folks who don't upgrade continue to use V1 interface and constructor. Folks who upgrade use the V2 interface and constructor.

V1 and V2 are now completely separate. I think this allows us to remove the inheritance of V1 from V2 and treat them completely separately. This has the benefit of V2 users not having to pseudo-implement the V1 interface. TreeWalkerSession code already works like this as well. Let me know what you think.

---
In reply to: [504869930](https://github.com/microsoft/Forge/pull/32#discussion_r504869930) [](ancestors = 504869930,504278218,504228085)

Reply:
cool. i like the idea to separate the two interfaces completely by removing TryUpgradeToITreeWalkerCallbacksV2().  Let me remove it and the associated test case.

---
In reply to: [505818977](https://github.com/microsoft/Forge/pull/32#discussion_r505818977) [](ancestors = 505818977,504869930,504278218,504228085)